Cannot compile retroarch on msys2 ucrt64.

### Is there an existing issue for this?

- [X] This is a bug in RetroArch frontend
- [X] I have searched the existing issues

### Description

Retro arch gets a syntax error because of new headers of mingw.

### Expected behavior

It should compile.


### Steps to reproduce the bug

install deps, git clone, configure, make.

### Version/Commit

839b7654dbf0fe75e1d6ffbac84f9b2ee79b51b5

### Bisect Results

_No response_

### Check in the nightly version

I don't know

### Platform & operating system

Windows 11

### Affected Cores

_No response_

### Environment information

MINGW64_NT-10.0-22631 windows 3.5.4-0bc1222b.x86_64 2024-12-05 09:27 UTC x86_64 Msys

### Relevant log output
<details><summary>Details</summary>
<p>

```shell
CC audio/drivers/xaudio.c
audio/drivers/xaudio.c: In function 'xaudio2_new':
audio/drivers/xaudio.c:279:127: error: macro "IXAudio2_CreateMasteringVoice" requires 8 arguments, but only 7 given
  279 |    if (FAILED(IXAudio2_CreateMasteringVoice(handle->pXAudio2, &handle->pMasterVoice, channels, samplerate, 0, idx_found, NULL)))
      |                                                                                                                               ^
In file included from audio/drivers/xaudio.h:21,
                 from audio/drivers/xaudio.c:39:
audio/drivers/xaudio29.h:617:9: note: macro "IXAudio2_CreateMasteringVoice" defined here
  617 | #define IXAudio2_CreateMasteringVoice(This,ppMasteringVoice,InputChannels,InputSampleRate,Flags,DeviceId,pEffectChain,StreamCategory) ((This)->lpVtbl->CreateMasteringVoice(This,ppMasteringVoice,InputChannels,InputSampleRate,Flags,DeviceId,pEffectChain,StreamCategory))
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from E:/msys64/ucrt64/include/winbase.h:2864,
                 from E:/msys64/ucrt64/include/windows.h:70,
                 from ./libretro-common/include/retro_miscellaneous.h:41,
                 from audio/drivers/xaudio.c:29:
audio/drivers/xaudio.c:279:15: error: 'IXAudio2_CreateMasteringVoice' undeclared (first use in this function); did you mean 'IXAudio2MasteringVoice'?
  279 |    if (FAILED(IXAudio2_CreateMasteringVoice(handle->pXAudio2, &handle->pMasterVoice, channels, samplerate, 0, idx_found, NULL)))
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
audio/drivers/xaudio.c:279:15: note: each undeclared identifier is reported only once for each function it appears in
audio/drivers/xaudio.c: In function 'xa_list_new':
audio/drivers/xaudio.c:497:4: error: implicit declaration of function 'IXAudio2_GetDeviceCount' [-Wimplicit-function-declaration]
  497 |    IXAudio2_GetDeviceCount(ixa2, &dev_count);
      |    ^~~~~~~~~~~~~~~~~~~~~~~
audio/drivers/xaudio.c:501:7: error: unknown type name 'XAUDIO2_DEVICE_DETAILS'; did you mean 'XAUDIO2_VOICE_DETAILS'?
  501 |       XAUDIO2_DEVICE_DETAILS dev_detail;
      |       ^~~~~~~~~~~~~~~~~~~~~~
      |       XAUDIO2_VOICE_DETAILS
audio/drivers/xaudio.c:502:11: error: implicit declaration of function 'IXAudio2_GetDeviceDetails'; did you mean 'IXAudio2Voice_GetVoiceDetails'? [-Wimplicit-function-declaration]
  502 |       if (IXAudio2_GetDeviceDetails(ixa2, i, &dev_detail) == S_OK)
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~
      |           IXAudio2Voice_GetVoiceDetails
audio/drivers/xaudio.c:504:59: error: request for member 'DisplayName' in something not a structure or union
  504 |          char *str = utf16_to_utf8_string_alloc(dev_detail.DisplayName);
      |                                                           ^
make: *** [Makefile:209: obj-unix/release/audio/drivers/xaudio.o] Error 1
```

</p>
</details> 


My assumption is that this wasn't defined in mingw yet, or that the mingw declarations changed, but I am not sure.

Trying mingw64 on win11 with latest msys2: (`pacman -Syu`)

```
audio/drivers/xaudio.c: In function 'xaudio2_new':
audio/drivers/xaudio.c:267:9: note: '#pragma message: The value of _WIN32_WINNT is: 0x0600'
  267 | #pragma message ("The value of _WIN32_WINNT is: " XSTR(_WIN32_WINNT))
      |         ^~~~~~~
audio/drivers/xaudio.c:283:127: error: macro "IXAudio2_CreateMasteringVoice" requires 8 arguments, but only 7 given
  283 |    if (FAILED(IXAudio2_CreateMasteringVoice(handle->pXAudio2, &handle->pMasterVoice, channels, samplerate, 0, idx_found, NULL)))
      |                                                                                                                               ^
In file included from audio/drivers/xaudio.h:21,
                 from audio/drivers/xaudio.c:39:
audio/drivers/xaudio29.h:617:9: note: macro "IXAudio2_CreateMasteringVoice" defined here
  617 | #define IXAudio2_CreateMasteringVoice(This,ppMasteringVoice,InputChannels,InputSampleRate,Flags,DeviceId,pEffectChain,StreamCategory) ((This)->lpVtbl->CreateMasteringVoice(This,ppMasteringVoice,InputChannels,InputSampleRate,Flags,DeviceId,pEffectChain,StreamCategory))
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
CC libretro-common/audio/conversion/float_to_s16.c
In file included from E:/msys64/mingw64/include/winbase.h:2864,
                 from E:/msys64/mingw64/include/windows.h:70,
                 from ./libretro-common/include/retro_miscellaneous.h:41,
                 from audio/drivers/xaudio.c:29:
audio/drivers/xaudio.c:283:15: error: 'IXAudio2_CreateMasteringVoice' undeclared (first use in this function); did you mean 'IXAudio2MasteringVoice'?
  283 |    if (FAILED(IXAudio2_CreateMasteringVoice(handle->pXAudio2, &handle->pMasterVoice, channels, samplerate, 0, idx_found, NULL)))
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
audio/drivers/xaudio.c:283:15: note: each undeclared identifier is reported only once for each function it appears in
audio/drivers/xaudio.c: In function 'xa_list_new':
audio/drivers/xaudio.c:501:4: error: implicit declaration of function 'IXAudio2_GetDeviceCount' [-Wimplicit-function-declaration]
  501 |    IXAudio2_GetDeviceCount(ixa2, &dev_count);
      |    ^~~~~~~~~~~~~~~~~~~~~~~
audio/drivers/xaudio.c:505:7: error: unknown type name 'XAUDIO2_DEVICE_DETAILS'; did you mean 'XAUDIO2_VOICE_DETAILS'?
  505 |       XAUDIO2_DEVICE_DETAILS dev_detail;
      |       ^~~~~~~~~~~~~~~~~~~~~~
      |       XAUDIO2_VOICE_DETAILS
audio/drivers/xaudio.c:506:11: error: implicit declaration of function 'IXAudio2_GetDeviceDetails'; did you mean 'IXAudio2Voice_GetVoiceDetails'? [-Wimplicit-function-declaration]
  506 |       if (IXAudio2_GetDeviceDetails(ixa2, i, &dev_detail) == S_OK)
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~
      |           IXAudio2Voice_GetVoiceDetails
audio/drivers/xaudio.c:508:59: error: request for member 'DisplayName' in something not a structure or union
  508 |          char *str = utf16_to_utf8_string_alloc(dev_detail.DisplayName);
      |                                                           ^
make: *** [Makefile:207: obj-unix/release/audio/drivers/xaudio.o] Error 1
make: *** Waiting for unfinished jobs....
CC libretro-common/audio/conversion/mono_to_stereo_float.c
```
qt was disabled.

--- a/.github/workflows/MSYS2.yml
+++ b/.github/workflows/MSYS2.yml
@@ -1,0 +1,77 @@
+name: CI Windows (MSYS2)
+
+on:
+  push:
+  pull_request:
+  repository_dispatch:
+    types: [run_build]
+
+permissions:
+  contents: read
+
+jobs:
+  msys2-build-test:
+    strategy:
+      fail-fast: false
+      matrix:
+        sys: [MINGW64, UCRT64,CLANG64]
+    runs-on: windows-latest
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Set up MSYS2
+        uses: msys2/setup-msys2@v2
+        with:
+          msystem: ${{ matrix.sys }}
+          update: true
+          install: base-devel git
+          pacboy: >-
+            gettext:p
+            gobject-introspection:p
+            graphite2:p
+            p11-kit:p
+            qt6:p
+            qt6-3d:p
+            qt6-charts:p
+            qt6-datavis3d:p
+            qt6-imageformats:p
+            qt6-location:p
+            qt6-lottie:p
+            qt6-networkauth:p
+            qt6-quick3dphysics:p
+            qt6-quicktimeline:p
+            qt6-remoteobjects:p
+            qt6-scxml:p
+            qt6-sensors:p
+            qt6-serialbus:p
+            qt6-speech:p
+            qt6-tools:p
+            qt6-translations:p
+            qt6-virtualkeyboard:p
+            qt6-webchannel:p
+            qt6-websockets:p
+            x264:p
+            cc:p
+
+      - name: Configure and build RetroArch
+        shell: msys2 {0}
+        run: |
+          echo "Building RetroArch in ${{ matrix.sys }} environment"
+          ./configure
+          make -j$(nproc)
+
+      - name: Collect DLLs and binaries
+        shell: msys2 {0}
+        run: |
+          echo "Collecting DLLs and binaries"
+          mkdir -p dist
+          cp retroarch.exe dist/
+          ldd retroarch.exe|grep $MINGW_PREFIX |awk '{print $3}'|xargs -I {} cp {} dist/
+
+      - name: Archive build artifacts
+        if: success()
+        uses: actions/upload-artifact@v4
+        with:
+          name: retroarch-${{ matrix.sys }}
+          path: dist/

--- a/Makefile.common
+++ b/Makefile.common
@@ -2648,4 +2648,32 @@ ifeq ($(HAVE_ODROIDGO2), 1)
          gfx/drivers/oga_gfx.o
 endif
 
+# Detect the operating system
+UNAME := $(shell uname -s)
+
+# Check if the system is MSYS2 (MINGW64 or MINGW32)
+ifneq ($(findstring MINGW,$(UNAME)),)
+$(info Detected MSYS2 environment)
+
+NT_VERSION := $(shell \
+  echo '#include <windows.h>' > temp.c; \
+  echo '#ifdef _WIN32_WINNT' >> temp.c; \
+  echo '#define GET_MACRO_VALUE(x) #x' >> temp.c; \
+  echo '#define EXPAND_MACRO_VALUE(x) GET_MACRO_VALUE(x)' >> temp.c; \
+  echo '#pragma message("_WIN32_WINNT=" EXPAND_MACRO_VALUE(_WIN32_WINNT))' >> temp.c; \
+  echo '#endif' >> temp.c; \
+  $(CC) -c temp.c 2>&1 | sed -n 's/^.*_WIN32_WINNT=\(0x[0-9A-Fa-f]\+\).*/\1/p'; \
+  rm -f temp.c temp.o)
+
+ifneq ($(NT_VERSION),)
+ifeq ($(shell [ $$(( $(NT_VERSION) )) -gt $$(( 0x602 )) ] && echo true),true)
+LIBS += -lxaudio2_9
+endif
+else
+$(warning Windows NT version macro (_WIN32_WINNT) is not defined.)
+endif
+
+
+endif
+
 ##################################

--- a/audio/common/mmdevice_common_inline.h
+++ b/audio/common/mmdevice_common_inline.h
@@ -20,10 +20,6 @@
 
 /* Fix for MSYS2 increasing _WIN32_WINNT to 0x0603*/
 #if defined(__MINGW32__) || defined(__MINGW64__)
-#ifdef _WIN32_WINNT
-#undef _WIN32_WINNT
-#endif
-#define _WIN32_WINNT 0x0600
 #define WIN32_LEAN_AND_MEAN
 #else
 typedef enum EDataFlow EDataFlow;

